src/core/indexer/mod.rs

```diff
@@ -0,0 +1,595 @@
+            if o != index && szg.os == 0 {
+                return err!(Error::InvalidVarIndex(format!(
+                    "Non matching ondex '{o}' and index '{index}' for code = '{code}'."
+                )));
+            }
+        }
+
+        // compute fs from index
+        let mut fs = szg.fs;
+        if fs == 0 {
+            if cs % 4 != 0 {
+                return err!(Error::InvalidCodeSize(format!(
+                    "Whole code size not multiple of 4 for variable length material. cs = '{cs}'."
+                )));
+            }
+
+            if szg.os != 0 {
+                return err!(Error::InvalidCodeSize(format!(
+                    "Non-zero other index size for variable length material. os = '{o}'.",
+                    o = szg.os
+                )));
+            }
+
+            fs = (index * 4) + cs
+        }
+
+        let rize = (fs - cs) * 3 / 4;
+        if raw.len() < rize as usize {
+            return err!(Error::Shortage(format!(
+                "insufficient raw material: raw size = '{}', rize = '{rize}'",
+                raw.len()
+            )));
+        }
+
+        Ok(Indexer { code, raw: raw[..rize as usize].to_vec(), index, ondex })
+    }
+
+    pub fn new_with_qb64(qb64: &str) -> Result<Self> {
+        let mut i: Indexer = Default::default();
+        i.exfil(qb64)?;
+        Ok(i)
+    }
+
+    pub fn new_with_qb64b(qb64b: &[u8]) -> Result<Self> {
+        let qb64 = String::from_utf8(qb64b.to_vec())?;
+
+        let mut i: Indexer = Default::default();
+        i.exfil(&qb64)?;
+        Ok(i)
+    }
+
+    pub fn new_with_qb2(qb2: &[u8]) -> Result<Self> {
+        let i: Indexer = Default::default();
+        i.bexfil(qb2)?;
+        Ok(i)
+    }
+
+    /// Extracts self.code, self.index, and self.raw from qualified base64 bytes qb64b
+    /// cs = hs + ss
+    /// ms = ss - os (main index size)
+    /// when fs None then size computed & fs = size * 4 + cs
+    fn exfil(&mut self, qb64: &str) -> Result<()> {
+        if qb64.is_empty() {
+            return err!(Error::EmptyMaterial("empty qb64".to_string()));
+        }
+
+        let first = qb64.chars().next().unwrap();
+        let hs = tables::hardage(first)? as usize;
+        if qb64.len() < hs {
+            return err!(Error::Shortage(format!(
+                "Need '{s}' more characters.",
+                s = (hs - qb64.len())
+            )));
+        }
+
+        let hard = &qb64[..hs];
+        let szg = tables::sizage(hard)?;
+
+        // both hard + soft code size
+        let cs = szg.hs + szg.ss;
+        let ms = szg.ss - szg.os;
+
+        if qb64.len() < cs as usize {
+            return err!(Error::Shortage(format!(
+                "Need '{l}' more characters",
+                l = ((cs as usize) - qb64.len()),
+            )));
+        }
+
+        let index = util::b64_to_u32(&qb64[hs..(hs + ms as usize)])?;
+        let odx = &qb64[(hs + ms as usize)..(hs + ms as usize + szg.os as usize)];
```

Comment:
You could reduce this slightly:
```suggestion
        let odx = &qb64[(hs + ms as usize)..(hs + (ms + szg.os) as usize)];
```